code/game/objects/items/glassjar.dm

```diff
@@ -120,13 +136,61 @@
 				underlays += victim
 				name = "glass jar with [S]"
 				desc = "A small jar with [S] inside."
+		if(JAR_GUMBALL)
+			name = "gumball jar"
+			desc = "A jar containing gumballs with varying colours."
+			var/image/gumballs_overlay = image(icon)
+			switch(length(gumballs_contained))
+				if(1 to GUMBALL_MIN)
+					gumballs_overlay.icon_state = "gumball_min"
+				if(6 to GUMBALL_MEDIUM)
+					gumballs_overlay.icon_state = "gumball_min"
+				if(11 to GUMBALL_MAX)
+					gumballs_overlay.icon_state = "gumball_max"
+			underlays += gumballs_overlay
+
 	return
 
+/obj/item/glass_jar/proc/handle_gumball_addition(mob/user, var/obj/item/clothing/mask/chewable/candy/gum/gumball/G) // handles the entire gumbball addition process
+	if(length(gumballs_contained) < GUMBALL_MAX)
+		gumballs_contained += G
+		user.drop_from_inventory(G)
+		G.forceMove(src)
+		if(!contains)
+			contains = JAR_GUMBALL
+		user.visible_message("[user] puts a gumball in \the [name].", SPAN_NOTICE("You put a gumball in \the [name]."))
+		handle_gumball_underlays()
+	else
+		to_chat(user, SPAN_WARNING("\The [name] is full!"))
+
+/obj/item/glass_jar/proc/handle_gumball_removal(mob/user) // same as the addition proc but we're removing gum instead
+	if(length(gumballs_contained))
+		user.put_in_hands(gumballs_contained[1])
+		gumballs_contained -= gumballs_contained[1]
+		user.visible_message("[user] takes a gumball from \the [name].", SPAN_NOTICE("You take a gumball from \the [name]."))
+		if(length(gumballs_contained) == 0)
+			contains = JAR_NOTHING
+		handle_gumball_underlays()
+
+/obj/item/glass_jar/proc/handle_gumball_underlays() // gumball overlays
+	underlays.Cut()
+	if(length(gumballs_contained))
+		var/image/gumballs_overlay = image(icon)
+		switch(length(gumballs_contained))
+			if(1 to GUMBALL_MIN)
+				gumballs_overlay.icon_state = "gumball_min"
+			if(6 to GUMBALL_MEDIUM)
+				gumballs_overlay.icon_state = "gumball_medium"
+			if(11 to GUMBALL_MAX)
+				gumballs_overlay.icon_state = "gumball_max"
+		underlays += gumballs_overlay
+
+
 /obj/item/glass_jar/peter/
 	name = "Peter's Jar"
 
 /obj/item/glass_jar/peter/Initialize()
-	. = ..()
+	..()
```

Comment:
you never add a return hint here, so it has to be `. = ..()` to return the old one